src/server.lisp

```diff
@@ -0,0 +1,184 @@
+    (loop
+      (handler-case
+          (let (request result reply)
+            (multiple-value-bind (identity empty-frame raw-request)
+                (%pull-raw-request receiver)
+              
+              (handler-case
+                  (progn
+                    (setf request (deserialize raw-request))
+                    (unless (typep request '|RPCRequest|)
+                      (error (make-instance 'not-an-RPCRequest
+                                            :object request)))
+                    (format-log logging-stream "Got request: ~a" (|RPCRequest-method| request))
+                    
+                    (let ((params-as-plist
+                            (loop :for key :being :the :hash-keys :of (|RPCRequest-params| request)
+                                  :with val := (gethash key (|RPCRequest-params| request))
+                                  :collect (str->lisp-keyword key)
+                                  :collect val)))
+                      (let ((f (gethash (|RPCRequest-method| request) dispatch-table)))
+                        (unless f
+                          (error (make-instance 'unknown-RPC-method
+                                                :method-name (|RPCRequest-method| request))))
+                        (setf result 
+                              (if timeout
+                                  (bt:with-timeout (timeout)
+                                    (apply f params-as-plist))
+                                  (apply f params-as-plist)))
+                        
+                        (setf reply (make-instance '|RPCReply|
+                                                   :|id| (|RPCRequest-id| request)
+                                                   :|result| result))))
+                    (format-log logging-stream "Finishing request: ~a" (|RPCRequest-method| request)))
+                
+                ;; this is where errors go where we can reply to the client
+                (unknown-RPC-method (c)
+                  (declare (ignore c))
+                  (format-log logging-stream "Error: method ~a unknown" (|RPCRequest-method| request))
+                  (setf reply (make-instance '|RPCError|
+                                             :|id| (|RPCRequest-id| request)
+                                             :|error| (format nil "Method named \"~a\" is unknown."
+                                                              (|RPCRequest-method| request)))))
+                (sb-ext:timeout (c)
+                  (declare (ignore c))
+                  (format-log logging-stream "Timed out on request ~a" (|RPCRequest-method| request))
+                  (setf reply (make-instance '|RPCError|
+                                             :|id| (|RPCRequest-id| request)
+                                             :|error| (format nil "Execution timed out.  Note: time limit: ~a seconds." timeout))))
+                (simple-error (c)
```

Comment:
Perhaps should be `error` instead of `simple-error`

Reply:
deffo, didn't know that was a thing